Added methods for retrieving filename and modifying file extensions

King2500 opened the pull request:
Coming from the C#/.NET world, the approciate "Path" class has a few more useful methods. These are only partially available in native PHP, so I thought this would be a good point to extend and complete your Path class with these:

```php
getFilename($path)
getFilenameWithoutExtension($path, $extension = null)
getExtension($path, $forceLowerCase = false)
hasExtension($path, $extension = null, $ignoreCase = false)
changeExtension($path, $extension)
```

I'm open for any discussion or if you don't like it / deny it.

webmozart commented on src/Path.php:
```diff
@@ -174,6 +175,132 @@ public static function getRoot($path)
+        }
+
+        return basename($path);
+	}
```
seems to be a tab here

webmozart commented on src/Path.php:
```diff
@@ -174,6 +175,132 @@ public static function getRoot($path)
+     *
+     * @param string $path  The path string
+     *
+     * @return string Filename.
```
Could you add

```
 * @return string Filename.
 *
 * @since 1.1
 */
```

here and in the other (non-test) methods?

webmozart commented on src/Path.php:
```diff
@@ -174,6 +175,132 @@ public static function getRoot($path)
+        }
+
+        if (null !== $extension) {
+            return trim(basename($path, $extension), '.');
```
is there a test for the trimming?

webmozart commented on src/Path.php:
```diff
@@ -174,6 +175,132 @@ public static function getRoot($path)
+        $ext = pathinfo($path, PATHINFO_EXTENSION);
+
+        if ($forceLowerCase) {
+            $ext = function_exists('mb_strtolower') ? mb_strtolower($ext) : strtolower($ext);
```
the test for the `mb_strtolower` case seems to be missing

webmozart commented on src/Path.php:
```diff
@@ -174,6 +175,132 @@ public static function getRoot($path)
+
+        // Make an array of extensions
+        if (!is_array($extension)) {
+            $extension = array(trim($extension, '.'));
```
is there a test for the trimming?

webmozart commented on src/Path.php:
```diff
@@ -174,6 +175,132 @@ public static function getRoot($path)
+
+        if ($ignoreCase) {
+            $strToLower = function_exists('mb_strtolower') ? 'mb_strtolower' : 'strtolower';
+            $extension = array_map($strToLower, $extension);
```
this block seems to be untested

webmozart commented on src/Path.php:
```diff
@@ -174,6 +175,132 @@ public static function getRoot($path)
+            return false;
+        }
+
+        $ext = self::getExtension($path, $ignoreCase);
```
can you rename `$ext` to `$actualExtension` here and in the other methods?

webmozart commented:
Well this is pretty cool! :) Thanks @King2500. I'll happily merge this when my comments are resolved

King2500 commented:
Thanks. :)
Added/changed the mentioned points and fixed little bugs (found through additional tests :))

I'm not sure about the requirement of mbstring for unit-testing.

Should we add
```json
    "require-dev": {
        "ext-mbstring": "*"
    }
```
to composor.json?
(Then my conditions in the test methods are obsolete)

webmozart commented:
I think the way you did it is fine. It's better to run 95% of the tests when mbstring is not installed than none at all. Thanks! :)
